### ISSUE TITLE
Redesign buffer class hierarchy 

### ISSUE THREAD

**Issue by Ambrevar** [atlas-engineer/nyxt#1805]
(I think we already had an issue about this but I can't seem to find it.  Please let me know if you do.)

Currently we have a top `buffer` class which contains way too much, and then a bunch of child classes.  Here is the approximate hierarchy:

```
- buffer
  - internal-buffer
    - prompt-buffer
    - panel-buffer 
    - editor-buffer
    - status-buffer
  - web-buffer
    - nosave-buffer
```

This design raises a few problems:
- proxy settings are missing for the prompt buffer
- https://github.com/atlas-engineer/nyxt/pull/1642: On FreeBSD the prompt buffer would take about 30s to show up because of a bug in the password manager library!

The last point is alarming: the prompt buffer should have nothing to do with it.  The prompt buffer performance issue arose because we do way too much in `buffer` and as such it becomes such a critical part of the code.  As Nyxt grows, we risk bankruptcy because of this.

Even `make-dummy-buffer` is much too expensive.

With @aartaka 's #1804 patch the buffer hierarchy is redesigned.  He thought we could merge everything into a `buffer` megaclass.  Indeed this simplifies the design (and thus avoid issue like prompt buffer proxy settings) but it does not address the bankruptcy issue.

So instead I suggest we start from a minimal `buffer` class, then use some mixin classes to still get a simple design.

```
- raw-buffer: almost no slots, not gtk-object; useful for dummy buffers.
  - proxied-buffer: proxy settings
  - buffer: gtk-object, url, title, modes, hooks, keymap, input, zoom... (nothing network-related)
    - internal-buffer
      - prompt-buffer (proxied-buffer)
      - panel-buffer (proxied-buffer)
      - ...
    - web-buffer (proxied-buffer): password manager, downloads, default-url, resource-request-hook...
      - nosave-buffer
```

Main question: should `buffer` have a web view?  The only buffers so far that don't need a web view are the dummy buffers.  In the above, I've introduced `raw-buffer` for this purpose.  I'm unsure about the naming, but introducing `raw-buffer` and keeping `buffer` has the benefit of introducing the least changes.

**Comment by Ambrevar** [comment]
@jmercouris @pdelfino What do you think?

**Comment by pdelfino** [comment]
This is a delicate and critical issue. I feel uneducated about this debate to be honest. 

Recently, I read the Sonja book on CLOS. Your suggestion with `buffer` as a **basic class** (author terminology, not technical consensus) and some mixin classes seems exactly what the author suggests for this kind of situation (if my interpretation of the book is correct).

Using Sonja's style for names, I guess the most basic class should be called `buffer`. Futher "descendants" would have names as variations of the basic class.

Thus, `raw-buffer` would be called `buffer`. This is problematic because in your model already exists a `buffer` class. Hence, this name would need to change too. Maybe changing the old `buffer` to `advanced-buffer`? Or `complex-buffer`?

My two cents :)

**Comment by aartaka** [comment]
`styleable-buffer` can be a useful mix-in too: both `internal-buffer` (or `buffer` in case of #1815) have `style` slot and we can benefit from one class to configure 'em all.

**Comment by Ambrevar** [comment]
Good call!

**Comment by aartaka** [comment]
> We also have `background-buffer`s now. @aartaka What do these buffers need? Just a gtk-object web-view, or also other slots?

They need everything `web-buffer` has, so that they can truthfully simulate its behavior.

> In particular, background buffers should not be added to the `buffer-list`.

Yes, that's the only difference, and it makes them invisible to the user.

**Comment by Ambrevar** [comment]
So should web buffers be children of background buffers then?

**Comment by aartaka** [comment]
> So should web buffers be children of background buffers then?

Not necessarily :) I believe it's still better to subclass `background-buffer` from `user-web-buffer`, so that all the changes from `web-buffer` configuration are propagated to `background-buffer`.

Making buffer visibility a configurable knob (or introducing methods for this?) can be a solution to this `background-buffer` uniqueness.

**Comment by Ambrevar** [comment]
True, but on the other hand class types and method give this for free if
web-buffer is a child of background buffer.

Note that all  changes would be propagated, because all the
slots would be in background-buffer, and web-buffer would be an empty
class.

We can also rename the classes if you like:
- web-buffer -> foreground-web-buffer
- background-buffer -> web-buffer

**Comment by Ambrevar** [comment]
But it's a nit, really... :)

**Comment by aartaka** [comment]
It doesn't feel right, but I can understand the benefits of this :)

**Comment by Ambrevar** [comment]
I'll have a go at this next.

**Comment by Ambrevar** [comment]
Fixed with #2224.
